_posts/2023-10-04-#28107.md

```diff
@@ -7,17 +7,55 @@ authors: [dergoegge]
+
+- In the context, of transaction identifiers, type-safety is important because it is impossible to deduce the type difference from just the `uint256` value. With `GenTxid` type checking can be done, but it is more verbose and done at run-time instead of at compile-time. Additionally, transaction identifier types improve readability of code by being explicit about expected types instead of solely relying on clear variable naming.
+
+- This PR aims to improve type safety by introducing the new [`transaction_identifier`](https://github.com/bitcoin-core-review-club/bitcoin/commit/28f9ff3d5416407d3c6d5374f6fc3c2767f4a7f3#diff-f3d91988bf7d98529d638fa829dc3fab695be002305a7614eaaeb283d4fb8101R18) type, with `Txid` and `Wtxid` typedefs for brevity.
+
+- There are instances where `wtxid` and `txid` are treated identically, such as for example in [`most_recent_block_txs`](https://github.com/bitcoin/bitcoin/blob/6619d6a8dca5a4d8e664a76526ac6bef3eb17831/src/net_processing.cpp#L1916-L1918), where the `uint256` type is used to handle this. In other contexts such as [`CTxMemPool::info`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/txmempool.cpp#L862) where `wtxid` and `txid` are treated differently but either is accepted, the [`GenTxid`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/primitives/transaction.h#L425) type is adopted.
+
 ## Questions
+
+### Concept
+
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+1. What does it mean for a transaction identifier to be type-safe? Why is that important or helpful? Are there any downsides?
+
+1. Conceptually, when writing new code that requires referring to transactions, when should you use `txid` and when should you use `wtxid`?
+
+1. In which concrete way(s) could using `transaction_identifier` instead of `uint256` help find existing bugs or prevent the introduction of new ones? On the other hand, could this change introduce new bugs?
+
+1. SegWit activated at block height 481,824. Do transactions in blocks < 481,824 also have a `wtxid`? For example, what about transaction with `txid` `dc01120a639283c11eeada6ce540178adfbbca7fe0d2f41076367cbc9e901822`?
+
+2. Are there any places where a (w)txid is returned as a `uint256` where it would not make sense to return a `transaction_identifier` instead?
+
+3. The [`GenTxid`](https://github.com/bitcoin/bitcoin/blob/dcfbf3c2107c3cb9d343ebfa0eee78278dea8d66/src/primitives/transaction.h#L425) class already exists. How does it already enforce type correctness, and how is it different from the approach in this PR?
+
+4. Are there any places where `GenTxid` is used but using `transaction_identifier` would be better?
+
+### Code Review
+
```

Comment:
Question idea: how is the type safety enforced? (Hint: what does the compiler say if you try to use a `Txid` in place of a `Wtxid` or vice versa?)

Reply:
I feel like this is already covered in

> Consider [`return m_wtxid_to_orphan_it.count(Wtxid::FromUint256(gtxid.GetHash()));`](https://github.com/bitcoin-core-review-club/bitcoin/commit/e74152d5b48bdca8dfc2217f35842dbb24b2746c#diff-7cfa3d46c37a17d80c78a37d1239b572a03095d50879ba610d87b702cd844f9dR177). Why is it not type-safe to allow `Wtxid` to be constructed by just passing a `uint256` to its constructor? Specifically, why doesn't `Wtxid{gtxid.GetHash()}` compile?